openhtmltopdf-core/src/main/java/com/openhtmltopdf/util/ThreadCtx.java

```diff
@@ -2,26 +2,64 @@
 
 import com.openhtmltopdf.layout.SharedContext;
 
+import java.io.Closeable;
+import java.util.function.Consumer;
+import java.util.logging.Level;
+
 /**
  * Because OpenHTMLtoPDF is designed to run in a single thread at all times for one invocation,
  * we can use a ThreadLocal to store pseudo global variables.
  * This MUST be set up in the appropriate renderer.
  */
 public class ThreadCtx {
-	private static final ThreadLocal<ThreadData> data = new ThreadLocal<ThreadCtx.ThreadData>() {
-		@Override
-		protected ThreadData initialValue() {
-			return new ThreadData();
-		}
-	};
-	
+
+	private static final ThreadLocal<ThreadData> data = ThreadLocal.withInitial(() -> new ThreadData());
+	private static final ThreadLocal<Consumer<Diagnostic>> diagnosticConsumer = new ThreadLocal();
+
+
+
 	public static ThreadData get() {
 		return data.get();
 	}
+
+	public static void addDiagnostic(String where, Level level, String diagnostic) {
+		addDiagnostic(where, level, diagnostic, null);
+	}
+
+	public static void addDiagnostic(String where, Level level, String msg, Throwable th) {
+		Consumer<Diagnostic> consumer = diagnosticConsumer.get();
+		if (consumer != null) {
+			consumer.accept(new Diagnostic(where, level, msg, th));
+		}
+	}
 	
 	public static void cleanup() {
 		data.remove();
 	}
+
+	public static Closeable applyDiagnosticConsumer(Consumer<Diagnostic> consumer) {
+		diagnosticConsumer.set(consumer);
+		return () -> diagnosticConsumer.remove();
+	}
+
+	public static class Diagnostic {
+		private final String where;
+		private final Level level;
+		private final String msg;
+		private final Throwable error;
+
+		public Diagnostic(String where, Level level, String msg, Throwable error) {
+			this.where = where;
+			this.level = level;
+			this.msg = msg;
```

Comment:
This implies that the user gets the message as string, wouldn't it be better to pass message enum and arg array in case they don't want to pay the price of formatting?

Reply:
Example: They are not interested in info level messages.

Reply:
you are right, it's something that I need to update. I'll switch to the enum+params.